internal/provider/resource_switch_port_settings.go

```
@@ -0,0 +1,1304 @@
+
+	// Save plan into Terraform state.
+	resp.Diagnostics.Append(resp.State.Set(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+}
+
+// Delete deletes the resource and removes the Terraform state on success.
+func (r *switchPortSettingsResource) Delete(ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse) {
+	// state is the resource data model for the current state.
+	var state switchPortSettingsModel
+
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	deleteTimeout, diags := state.Timeouts.Delete(ctx, defaultTimeout())
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	_, cancel := context.WithTimeout(ctx, deleteTimeout)
+	defer cancel()
+
+	if err := r.client.NetworkingSwitchPortSettingsDelete(
+		ctx,
+		oxide.NetworkingSwitchPortSettingsDeleteParams{
+			PortSettings: oxide.NameOrId(state.ID.ValueString()),
+		}); err != nil {
+		if !is404(err) {
+			resp.Diagnostics.AddError(
+				"Error deleting Switch Port Settings:",
+				"API error: "+err.Error(),
+			)
+			return
+		}
+	}
+	tflog.Trace(ctx, fmt.Sprintf("deleted Switch Port Settings with ID: %v", state.ID.ValueString()), map[string]any{"success": true})
+
+}
+
+func toTerraformModel(settings *oxide.SwitchPortSettings) (switchPortSettingsModel, diag.Diagnostics) {
+	var diags diag.Diagnostics
+
+	model := switchPortSettingsModel{
+		ID:          types.StringValue(settings.Id),
+		Name:        types.StringValue(string(settings.Name)),
+		Description: types.StringValue(settings.Description),
+		PortConfig: &switchPortSettingsPortConfigModel{
+			Geometry: types.StringValue(string(settings.Port.Geometry)),
+		},
+		TimeCreated:  types.StringValue(settings.TimeCreated.String()),
+		TimeModified: types.StringValue(settings.TimeModified.String()),
+	}
+
+	//
+	// Addresses
+	//
+	if len(settings.Addresses) > 0 {
```

`Addresses` is a non-nullable [required](https://github.com/oxidecomputer/omicron/blob/75d9f46a44ea2c0d94ca666016ca44d1f89fb18a/openapi/nexus.json#L24683) field in the API. It would be surprising (a misbehaviour the user should be aware of) if we received an empty list no? Maybe we should return an error if `len(settings.Addresses) < 1`? I'm not sure, WDYT? 🤔

---

I'll comment this further in the code. This was a bit annoying to deal with. The short answer is without this change Terraform will have a non-empty refresh plan.

The long answer is upon create Terraform will set the state for these attributes to `[]`, assuming they are unset in the Terraform configuration. Then a subsequent plan will see the state as `[]` but the configuration as `null` and trigger an `Update` call to correct that. Then the `Read` will set it back to `[]` and the cycle continues.

You're correct that the Oxide API will sent these values back as `[]` since they are required but in Terraform there's a distinction between null and `[]`. I chose to treat everything as null to avoid non-empty refresh plans for a configuration that was just applied. If I were to treat everything as `[]`, I'd have to enter a lot of defaults for the optional attributes, which is unwieldy for some of the larger attributes (e.g., `bgp_peers`).